### ISSUE TITLE
Bug: semi: "never" behaves differently since 9.23.0

### ISSUE THREAD

**Issue by woeye** [eslint/eslint#19575]
### Environment

Node version: 22.14.0
npm version: 10.9.2
Local ESLint version: 9.23.0
Global ESLint version:
Operating System: macOS Sequoia 15.3.2

### What parser are you using?

Default (Espree)

### What did you do?

After upgrading to v9.23.0 I noticed that all semicolons at the end of a function no longer were reported as an error, even though it should be according to my rules:

```js
const eslintConfig = [
  ...compat.extends('next/core-web-vitals', 'next/typescript'),
  {
    rules: {
      'semi': ['error', 'never'],
      'quotes': ['error', 'single'],
    }
  }
]
```

Downgrading back to v9.22.0 fixes the issue.

### What did you expect to happen?

If the rule semi is set to "never" it also should report semicolons at the end of a function as an error.

### What actually happened?

From the example above: the first line will be correctly reported as an error. However the semicolon at the end of the function will not.

Semicolon at the end of function is no longer reported as an error.

### Link to Minimal Reproducible Example

https://gist.github.com/woeye/6b24f61060fa3d5fa3768e22a7649f4a

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

This works fine with v9.22.0

**Comment by woeye** [comment]
Thanks for your response! Indeed, the "no-extra-semi" appears to solve the issue. I think I never noticed this before because normally I would write it like so:

```js
const cn = function(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
};

export { cn };
```

In this case both `;` are handled by the semi rule. 

With that said I would say the ticket can be closed as a non-bug. Sorry for the confusion.

**Comment by mdjermanovic** [comment]
Thanks for the info, I'm closing the issue then.
